File: js/0163-missingRanges.js
//	Time: 	 O(n)
//	Space:	 O(n)
/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {string[]}
 */
var findMissingRanges = function(nums, lower, upper) {
  const ranges = [];
  const upperNums = nums.concat([upper + 1])

  for (const n of upperNums) {
    if (n > lower) {
      const end = n - 1;
      const range = end === lower ? `${lower}` : `${lower}->${end}`;
      ranges.push(range);
    }

    lower = n + 1;
  }

  return ranges;
};

console.log(findMissingRanges([1,3,5,7,9], 0, 10));

File: py/0163-Missing_Ranges.py
#	Time: 	 O()
#	Space:	 O()

#############################################
#                                           #
#   Clever solution from discussion board   #
#                                           #
#############################################
# def stringify(a,b):
# 		if a == b:
# 			return str(a)
# 		else:
# 			return str(a) + "->" + str(b)

# 	nums = [lower-1] + nums + [upper+1] # that's the trick to avoid edge cases

# 	ans = [] 
# 	for i in range(1,len(nums)):
# 		if nums[i] - nums[i-1] > 1:
# 			ans.append(stringify(nums[i-1]+1,nums[i]-1))
# 	return ans